src/core/thread/mle.cpp

```diff
@@ -4743,5 +4769,72 @@ void Mle::DelayedResponseMetadata::RemoveFrom(Message &aMessage) const
     SuccessOrAssert(aMessage.SetLength(aMessage.GetLength() - sizeof(*this)));
 }
 
+Error Mle::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(!IsDetachingGracefully(), error = kErrorBusy);
+
+    OT_ASSERT(mDetachGracefullyCallback == nullptr);
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    if (IsChild() || IsRouter())
+    {
+        mDetachGracefullyTimer.Start(kDetachGracefullyTimeout);
+    }
+    else
+    {
+        mDetachGracefullyTimer.Start(0);
+    }
```

Comment:
May be even better to use `IsAttached`. We can simplify this using `? :`:
```suggestion
    mDetachGracefullyTimer.Start(IsAttached() ? kDetachGracefullyTimeout : 0);
```

Here is its definition: 
```c++
bool Mle::IsAttached(void) const
{
    return (IsChild() || IsRouter() || IsLeader());
}
```

Reply:
May also help to add a comment on why we start the timer with zero here:
(I know I may not remember if I read the code again later 😄 )

```
// If already detached or disabled, we start timer with zero duration so to 
// stop and invoke the callback when timer fires.
```

Reply:
Added comment to explain the 0 timer.

I didn't change to use `IsAttached` as we don't include the leader state here.